File: internal/database/drivers/mysql/role.go
package driver_mysql

import (
	"fmt"

	"github.com/paularlott/knot/internal/database/model"

	_ "github.com/go-sql-driver/mysql"
)

func (db *MySQLDriver) SaveRole(role *model.Role) error {

	tx, err := db.connection.Begin()
	if err != nil {
		return err
	}

	// Test if the PK exists in the database
	var doUpdate bool
	err = tx.QueryRow("SELECT EXISTS(SELECT 1 FROM roles WHERE role_id=?)", role.Id).Scan(&doUpdate)
	if err != nil {
		tx.Rollback()
		return err
	}

	// Update
	if doUpdate {
		err = db.update("roles", role, nil)
	} else {
		err = db.create("roles", role)
	}
	if err != nil {
		tx.Rollback()
		return err
	}

	tx.Commit()

	return nil
}

func (db *MySQLDriver) DeleteRole(role *model.Role) error {
	_, err := db.connection.Exec("DELETE FROM roles WHERE role_id = ?", role.Id)
	return err
}

func (db *MySQLDriver) GetRole(id string) (*model.Role, error) {
	var roles []*model.Role

	err := db.read("roles", &roles, nil, "role_id = ?", id)
	if err != nil {
		return nil, err
	}

	if len(roles) == 0 {
		return nil, fmt.Errorf("user role not found")
	}

	return roles[0], nil
}

func (db *MySQLDriver) GetRoles() ([]*model.Role, error) {
	var roles []*model.Role

	err := db.read("roles", &roles, nil, "1 ORDER BY name")
	if err != nil {
		return nil, err
	}

	return roles, nil
}

File: script-apiclient/lib/knot/role.py
# knot.role - Role management library for Knot server

from . import api

def list():
    """List all roles."""
    response = api.get("/api/roles")

    result = []
    for role in response.get("roles", []):
        result.append({
            "id": role.get("role_id"),
            "name": role.get("name")
        })

    return result


def get(role_id):
    """Get role by ID."""
    response = api.get(f"/api/roles/{role_id}")

    return {
        "id": response.get("role_id"),
        "name": response.get("name"),
        "permissions": response.get("permissions", [])
    }


def create(name, permissions=None):
    """Create a new role."""
    body = {
        "name": name,
        "permissions": permissions or []
    }

    response = api.post("/api/roles", body)
    return response.get("role_id")


def update(role_id, name=None, permissions=None):
    """Update role properties."""
    current = api.get(f"/api/roles/{role_id}")

    body = {
        "name": name if name is not None else current.get("name"),
        "permissions": permissions if permissions is not None else current.get("permissions", [])
    }

    api.put(f"/api/roles/{role_id}", body)
    return True


def delete(role_id):
    """Delete a role."""
    api.delete(f"/api/roles/{role_id}")
    return True
